Zend/zend_compile.c

```diff
@@ -5340,67 +5354,53 @@ void zend_compile_params(zend_ast *ast, zend_ast *return_type_ast) /* {{{ */
 		arg_info->type = ZEND_TYPE_ENCODE(0, 1);
 
 		if (type_ast) {
-			zend_bool has_null_default = default_ast
-				&& (Z_TYPE(default_node.u.constant) == IS_NULL
-					|| (Z_TYPE(default_node.u.constant) == IS_CONSTANT_AST
-						&& Z_ASTVAL(default_node.u.constant)->kind == ZEND_AST_CONSTANT
-						&& strcasecmp(ZSTR_VAL(zend_ast_get_constant_name(Z_ASTVAL(default_node.u.constant))), "NULL") == 0));
+			zend_bool has_null_default = default_ast && zend_is_null_const(default_node.u.constant);
 
 			op_array->fn_flags |= ZEND_ACC_HAS_TYPE_HINTS;
 			arg_info->type = zend_compile_typename(type_ast, has_null_default);
 
 			if (ZEND_TYPE_CODE(arg_info->type) == IS_VOID) {
 				zend_error_noreturn(E_COMPILE_ERROR, "void cannot be used as a parameter type");
 			}
-
-			if (type_ast->kind == ZEND_AST_TYPE) {
-				if (ZEND_TYPE_CODE(arg_info->type) == IS_ARRAY) {
-					if (default_ast && !has_null_default
-						&& Z_TYPE(default_node.u.constant) != IS_ARRAY
-						&& Z_TYPE(default_node.u.constant) != IS_CONSTANT_AST
-					) {
+			if(default_ast && !has_null_default && Z_TYPE(default_node.u.constant) != IS_CONSTANT_AST){
+				if (type_ast->kind == ZEND_AST_TYPE) {
+					if (ZEND_TYPE_CODE(arg_info->type) == IS_ARRAY && Z_TYPE(default_node.u.constant) != IS_ARRAY) {
```

Comment:
It should be possible to move the IS_ARRAY/IS_CALLABLE checks into the ZEND_TYPE_CODE switch below.